core/src/worker/tuner/resource_based.rs

```diff
@@ -768,4 +915,204 @@ mod tests {
+
+        let cgroup_info = CGroupCpuInfo::new(fake_fs);
+
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+
+        // No additional usage -> percentage should resolve to exactly zero.
+        assert_eq!(Some(0.0), cgroup_info.calc_cpu_percent());
+
+        // Add some usage such that it we should be above 0, but < 1
+        stat.replace(Some("usage_usec 1000".into()));
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+
+        let cpu_percent = cgroup_info
+            .calc_cpu_percent()
+            .expect("expected usage value");
+        assert!(
+            cpu_percent > 0. && cpu_percent < 1.0,
+            "epected cpu usage between (0.0, 1.0)"
+        );
+
+        // Simulate a large usage increase that should saturate the quota.
+        stat.replace(Some("usage_usec 1000000".into()));
+
+        let cpu_percent = cgroup_info
+            .calc_cpu_percent()
+            .expect("expected usage value");
+        assert!(
+            cpu_percent > 1.0,
+            "expected cpu usage greater than 1.0, got {cpu_percent}"
+        );
+    }
+
+    #[test]
+    fn cgroup_unlimited_quota_is_ignored() {
+        let stat = Rc::new(RefCell::new(Some("usage_usec 1000".into())));
+        let limit = Rc::new(RefCell::new(Some("max 1000".into())));
+
+        let cgroup_info = CGroupCpuInfo::new(FakeCGroupFS::new(stat.clone(), limit.clone()));
+
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+
+        // Unlimited quota is treated as disabled regardless of usage.
+        assert!(cgroup_info.calc_cpu_percent().is_none());
+        stat.replace(Some("usage_usec 2000".into()));
+
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+        assert!(cgroup_info.calc_cpu_percent().is_none());
+    }
+
+    #[test]
+    fn cgroup_stat_file_temporarily_unavailable() {
+        let stat = Rc::new(RefCell::new(None));
+        let limit = Rc::new(RefCell::new(Some("1000 1000".into())));
+
+        let cgroup_info = CGroupCpuInfo::new(FakeCGroupFS::new(stat.clone(), limit.clone()));
+
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+
+        assert!(cgroup_info.calc_cpu_percent().is_none());
+
+        // When available, the next call finishes initialization
+        stat.replace(Some("usage_usec 1000".into()));
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+        assert!(cgroup_info.calc_cpu_percent().is_none());
+
+        // A third call properly calculates percentage
+        stat.replace(Some("usage_usec 1500".into()));
+        std::thread::sleep(Duration::from_micros(1_000_u64));
+        let cpu_percent = cgroup_info
+            .calc_cpu_percent()
+            .expect("expected usage value");
+        assert!(cpu_percent > 0., "expected cpu percent > 0");
+    }
+
+    #[test]
+    fn cgroup_realsysinfo_uses_cgroup_limits_cpu() {
+        let sys_info = RealSysInfo::new();
+        let cgroup_info = CGroupCpuInfo::new(CgroupV2CpuFileSystem);
```

Comment:
Honestly this is one of the few spots where I don't like Rust - really this should be an integ test. It could be its own binary and only run in CI. Having to have the skips in a unit test feels a little awkward, but having to `pub` all this or put it in its own crate is even stupider.

I don't have any actual suggestion here, just whining.

... Actually maybe I do - let's protect these with explicit env vars saying to run the tests. That way they also won't silently succeed when they're _supposed_ to be able to read stuff but can't